subsys/ipc/ipc_service/backends/ipc_icmsg.c

```diff
@@ -43,10 +43,64 @@ static int send(const struct device *instance, void *token,
 	return icmsg_send(conf, dev_data, msg, len);
 }
 
+static int get_tx_buffer(const struct device *instance, void *token,
+			 void **data, uint32_t *user_len, k_timeout_t wait)
+{
+	const struct icmsg_config_t *conf = instance->config;
+	struct icmsg_data_t *dev_data = instance->data;
+
+	return icmsg_get_tx_buffer(conf, dev_data, data,
+				      user_len);
+}
+
+static int drop_tx_buffer(const struct device *instance, void *token,
+			  const void *data)
+{
+	const struct icmsg_config_t *conf = instance->config;
+	struct icmsg_data_t *dev_data = instance->data;
+
+	return icmsg_drop_tx_buffer(conf, dev_data, data);
+}
+
+static int send_nocopy(const struct device *instance, void *token,
+			const void *data, size_t len)
+{
+	const struct icmsg_config_t *conf = instance->config;
+	struct icmsg_data_t *dev_data = instance->data;
+
+	return icmsg_send_nocopy(conf, dev_data,
+				    data, len);
+}
+
+#ifdef CONFIG_IPC_SERVICE_ICMSG_NOCOPY_RX
+int hold_rx_buffer(const struct device *instance, void *token, void *data)
+{
+	const struct icmsg_config_t *conf = instance->config;
+	struct icmsg_data_t *dev_data = instance->data;
+
+	return icmsg_hold_rx_buffer(conf, dev_data, data);
+}
+
+int release_rx_buffer(const struct device *instance, void *token, void *data)
+{
+	const struct icmsg_config_t *conf = instance->config;
+	struct icmsg_data_t *dev_data = instance->data;
+
+	return icmsg_release_rx_buffer(conf, dev_data, data);
+}
+#endif /* CONFIG_IPC_SERVICE_ICMSG_NOCOPY_RX */
+
 const static struct ipc_service_backend backend_ops = {
 	.register_endpoint = register_ept,
 	.deregister_endpoint = deregister_ept,
 	.send = send,
+	.get_tx_buffer = get_tx_buffer,
+	.drop_tx_buffer = drop_tx_buffer,
+	.send_nocopy = send_nocopy,
+#ifdef CONFIG_IPC_SERVICE_ICMSG_NOCOPY_RX
```

Comment:
I think it would be cleaner if instead of relying on a icmsg library configuration (`IPC_SERVICE_ICMSG_NOCOPY_RX`) you create a new `IPC_SERVICE_BACKEND_ICMSG_NOCOPY_RX` in a new file like ipc_service/backends/Kconfig.icmsg to configure the icmsg backend.